crates/blockchain/blockchain.rs

```diff
@@ -340,128 +342,145 @@ impl Blockchain {
         for i in 0..16 {
             let (tx, rx) = channel();
             let handle = std::thread::Builder::new()
                 .name(format!("block_executor_merkleization_shard_worker_{i}"))
                 .spawn_scoped(scope, move || {
-                    self.handle_merkleization_subtrie(rx, parent_header)
+                    self.handle_merkleization_subtrie(rx, parent_header, i)
                 })
                 .map_err(|e| StoreError::Custom(format!("spawn failed: {e:?}",)))?;
             workers_handles.push(handle);
             workers_tx.push(tx);
         }
-        let mut state_updates_map: FxHashMap<Nibbles, Vec<u8>> = Default::default();
-        let mut storage_updates_map: StoreUpdatesMap = Default::default();
-        let mut code_updates: FxHashMap<H256, Code> = Default::default();
-        let mut hashed_address_cache: FxHashMap<H160, H256> = Default::default();
+
+        let mut account_updates = Vec::new();
+        let mut account_state: FxHashMap<H256, PreMerkelizedAccountState> = Default::default();
+        let mut code_updates: Vec<(H256, Code)> = vec![];
+
         for updates in rx {
             let current_length = queue_length.fetch_sub(1, Ordering::Acquire);
             *max_queue_length = current_length.max(*max_queue_length);
-            let mut hashed_updates: Vec<_> = updates
-                .into_iter()
-                .map(|u| {
-                    let hashed_address = hashed_address_cache
-                        .entry(u.address)
-                        .or_insert_with(|| keccak(u.address));
-                    (*hashed_address, u)
-                })
-                .collect();
-            hashed_updates.sort_by_key(|(h, _)| h.0[0]);
-            for sharded_update in hashed_updates.chunk_by(|l, r| l.0.0[0] & 0xf0 == r.0.0[0] & 0xf0)
-            {
-                let shard_message = sharded_update.to_vec();
-                workers_tx[(shard_message[0].0.0[0] >> 4) as usize]
-                    .send(shard_message)
-                    .map_err(|e| StoreError::Custom(format!("send failed: {e}")))?;
+            for update in updates {
+                let hashed_address = keccak(update.address);
+                let account_bucket = hashed_address.as_fixed_bytes()[0] >> 4;
+                workers_tx[account_bucket as usize]
+                    .send(MerklizationRequest::LoadAccount(hashed_address))
+                    .map_err(|e| StoreError::Custom(format!("send error: {e}")))?;
+
+                account_updates.push(update.clone());
+                if update.removed_storage | update.removed {
+                    for tx in &workers_tx {
+                        tx.send(MerklizationRequest::Delete(hashed_address))
+                            .map_err(|e| StoreError::Custom(format!("send error: {e}")))?;
+                    }
+                }
+                for (key, value) in update.added_storage {
+                    let hashed_key = keccak(key);
```

Comment:
We should check if all this hashing doesn't hurt performance. One alternative is to use a cheaper hash here, and leave the keccak to the workers.